spec.md

```diff
@@ -0,0 +1,954 @@
+    // This field is OPTIONAL.
+    map<string, bytes> Secrets = 2 [(cmi_secret) = true];
+}
+
+message ListMachinesResponse {
+    // MachineList is the map of list of machines. Format for the map should be map<MachineID, MachineName>.
+    // This field is REQUIRED.
+    map<string, string> MachineList = 1;
+}
+```
+
+##### ListMachines Errors
+
+If the plugin is unable to complete the ListMachines call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CMI-Client MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+| gRPC Code | Condition | Description | Recovery Behavior | Auto Retry Required |
+|-----------|-----------|-------------|-------------------|------------|
+| 0 OK | Successful | The call for listing all VMs associated with `ProviderSpec` was successful. |  | N |
+| 1 CANCELED | Cancelled | Call was cancelled. Perform any pending clean-up tasks and return the call |  | N |
+| 2 UNKNOWN | Something went wrong | Not enough information on what went wrong | Retry operation after sometime | Y |
+| 3 INVALID_ARGUMENT | Re-check supplied parameters | Re-check the supplied `ProviderSpec` and make sure that all required fields are present in their desired value format. Exact issue to be given in `.message` | Update `ProviderSpec` to fix issues. | N |
+| 4 DEADLINE_EXCEEDED | Timeout | The call processing exceeded supplied deadline | Retry operation after sometime | Y |
+| 5 NOT_FOUND |  |  |  |  |
+| 6 ALREADY_EXISTS |  |  |  |  |
+| 7 PERMISSION_DENIED | Insufficent permissions | The requestor doesn't have enough permissions to list VMs and it's required dependencies | Update requestor permissions to grant the same | N |
+| 8 RESOURCE_EXHAUSTED |  |  |  |  |
+| 9 PRECONDITION_FAILED |  |  |  |  |
+| 10 ABORTED |  |  |  |  |
+| 11 OUT_OF_RANGE |  |  |  |  |  |
+| 12 UNIMPLEMENTED | Not implemented | Unimplemented indicates operation is not implemented or not supported/enabled in this service. | Retry with an alternate logic or implement this method at the plugin. Most methods by default are in this state | N |
+| 13 INTERNAL | Major error | Means some invariants expected by underlying system has been broken. If you see one of these errors, something is very broken. | Needs manual intervension to fix this | N |
+| 14 UNAVAILABLE | Not Available | Unavailable indicates the service is currently unavailable. | Retry operation after sometime | Y |
+| 15 DATALOSS |  |  |  |  |
+| 16 UNAUTHENTICATED | Missing provider credentials | Request does not have valid authentication credentials for the operation | Fix the provider credentials | N |
+
+The status `message` MUST contain a human readable description of error, if the status `code` is not `OK`.
+This string MAY be surfaced by CMI-Client to end users.
+
+#### `GetListOfVolumeIDsForExistingPVs`
```

Comment:
This naming is somewhat clumsy IMO. Perhaps you could use a shorter name, e.g. `GetVolumeIDs` or at least `GetVolumeIDsForExistingPVs`? Also, a sentence about why such a RPC call is useful, e.g. for what purpose it could be used by a client would be great, at least to me it's not entirely obvious.